### ISSUE TITLE
Explain the *Observer pattern, especially when it is and isn't appropriate

### ISSUE THREAD

**Issue by domenic** [w3ctag/design-principles#78]
I've been seeing a recent trend toward XObservers: IntersectionObserver, IndexedDBObserver, ResizeObserver, etc.

We should explain how this pattern canonically works (see e.g. the discussion in https://github.com/WICG/indexed-db-observers/issues/43 about naming). But, more importantly, we should explain when it is and isn't appropriate. XObservers are *not* the new EventTarget, and should not be.

By default, you should be using events. So, when is an observer needed? I can think of some ideas but I'm not sure they're conclusive:

- When you want batching (but, events can batch too... this is not a great reason...)
- When you need to heavily customize the observing process with lots of options (thus need a constructor to take those options)
- When you might reuse the same observer callback on multiple different targets

Pinging various XObserver authors here for their thoughts: @annevk, @atotic, @dmurph, @szager-chromium.

**Comment by surma** [comment]
As a potential additional criterium for Observers: What about idle time dispatch á la IntersectionObserver. Or is that a thing that events can/should do as well? Is there precedent?

**Comment by annevk** [comment]
I suspect @smaug---- would be better to ask given he worked on mutation observers more closely.

Some reasons I can think of:

1. Mutation events dispatched synchronously and that was bad for various reasons.
2. It was too easy to observe too much with events and slow the whole thing down.

(Also, for Fetch we're considering a `FetchObserver` that's an `EventTarget`... I guess we could rename it, but it's a pretty apt name.)

**Comment by annevk** [comment]
@surma anything you can do with callbacks you can do with events as well. It ends up being more verbose if you just need one type though, but after that events tend to become worth it.

**Comment by smaug----** [comment]
I think the initial comment here listed the reasons quite well why the pattern was used/created for MutationObserver (IIRC I initially even used the name MutationBatch).
It is basically the combination of those 3 items: batching, options (either via ctor or observe() method, depending on the API), and multiple observables.

> What about idle time dispatch á la IntersectionObserver.

FWIW, IntersectionObserver doesn't use idle time.
And MutationObserver uses microtask.
So perhaps a common thing is that observing is asynchronous from JS point of view.
(but not necessarily from browser point of view, since microtasks are kind of sync.)

**Comment by szager-chromium** [comment]
In my opinion, the biggest difference between observers and events is this:

Observers give you controls to filter out events in which you're not interested.  For example, you could do something exactly functionally equivalent to IntersectionObserver by using scroll event handlers.  The reason -- the *only* reason -- we discourage that approach is because the scroll event handler winds up running *much* more often than necessary.

**Comment by atotic** [comment]
The main reason ResizeObserver uses "observer" pattern is performance. DOM event propagation (capture/bubble) can be costly, and ResizeObserver notification delivery loop is on a critical rendering path.

"scroll" event does not use standard event propagation rules for performance, and developers get [confused](https://stackoverflow.com/questions/37545308/scroll-event-isnt-propagated-to-parent).

I'd add:

* When observer is unable to pay the performance cost of event propagation.

**Comment by atotic** [comment]
> When you might reuse the same observer callback on multiple different targets

Is this is a valid pro-observer argument?  You could just register same callback on multiple elements.

**Comment by annevk** [comment]
> When broadcaster is unable to pay the performance cost of event propagation.

That only really applies if there's a propagation path. But that's also solved with `interface SpecialSnowFlake : EventTarget { ... };`.

**Comment by jakearchibald** [comment]
In terms of propagation & applying filters, that could be solved by having a method that returns an event target. But maybe "observers" are in place of that pattern.

The most compelling reason is asynchronicity. The results of the action are observable, but the callback hasn't been called yet.

**Comment by szager-chromium** [comment]
Not all observers are asynchronous (e.g. ResizeObserver).

**Comment by domenic** [comment]
We need to bust the myth that observers have something to do with asynchronicity/timing/delaying/batching. Events can be fired at any time, not just synchronously and immediately. You need a better reason than that for using an XObserver.

**Comment by jakearchibald** [comment]
@szager-chromium 

> Not all observers are asynchronous (e.g. ResizeObserver).

I think they should be considered async. See the console http://output.jsbin.com/sohufoq/quiet.

@domenic 

> We need to bust the myth that observers have something to do with asynchronicity

Do you feel the same about "the results of an action shouldn't be script-observable before its related event".

For example, with resize observers I can tell that the element has changed size before the callback fires, so it should be an observer, not an event. Same goes for mutation observers, I can see the DOM is different before the callback fires.

Whereas with [`AbortSignal`](https://dom.spec.whatwg.org/#interface-AbortSignal), the `.aborted` property can't be observed to be true before the related event.

**Comment by szager-chromium** [comment]
ResizeObserver, as spec'ed, is most definitely synchronous.  It should not be considered async, because it is not async.

**Comment by annevk** [comment]
@jakearchibald I don't think that matters. When `input` is dispatched you can also have observed that the input has changed. Again, events are just fancy callbacks. Observers seem to center mostly around wanting a single callback with lots of customization around it. I don't think we should read much into any of the other aspects as those depend on what you want to achieve.

**Comment by jakearchibald** [comment]
@szager-chromium maybe we have different definitions of async, but ResizeObserver's event is dispatched as part of the render steps of the event loop. The event loop can iterate many times before the next render steps.

My demo shows that happening. The browser's layout model updates without triggering the observer. Also, another queued task happens before the observer triggers, because many tasks can be handled between render steps.

It's as synchronous as `requestAnimationFrame`.

**Comment by dmurph** [comment]
Also adding @bsittler who is working on AsyncCookies which is an example of an event-based model that does batching and has a lot of options. This needs to start a serviceworker that had previously registered interest in a specific cookie pattern and tell it the cookie changed, which fits really well with dispatching an event on that serviceworker.

I think the main reason IDB Observers, and probably intersection observer, uses callbacks is because it already needs a lot of configuration around lifetime. Both involve a target being observed that can both die before or outlive the observer object & callback. Since that configuration involves creating an observer object, it was easier to just have a callback for the observer.

**Comment by szager-chromium** [comment]
@jakearchibald Yes, I see your point, but we need to be very careful in explaining this, because different XObserver types still have different timing of callbacks.  ResizeObserver, as you say, is effectively rAF timing, but IntersectionObserver is NOT.  By the time IntersectionObserver callbacks run, the frame has already been displayed, so it cannot be used to drive frame-precise layout.

So, it's generally true that observer notifications are delivered "after the fact", but I don't think we can say anything more specific than that.

**Comment by szager-chromium** [comment]
@domenic I don't agree with the premise that "by default, you should use events."  In fact, one of the primary motivations behind IntersectionObserver was to reduce the abuse of scroll event handlers.

I might actually go so far as to invert that logic: by default, you should use XObservers.  Only use event handlers if an observer isn't granular enough.

**Comment by domenic** [comment]
XObservers are un-idiomatic and heavy-weight. They involve creating a new class, options dictionary, and then also a type (sometimes a dictionary, sometimes a class) for the delivered records. They do not fit with any existing systems that play well with events (e.g. jQuery, RxJS, ...) and they do not benefit from any infrastructure changes we make to EventTarget in the future (e.g. adding `once` or other event listener options).

XObservers have their uses, but they should be a last resort, when events have proven insufficient.

It sounds like there was a problem with scroll events that made you use IntersectionObserver, but that sounds specific to scroll events (and perhaps specific to how they were specced and implemented, instead of to the concept in general).

**Comment by szager-chromium** [comment]
As far as I can tell, only IntersectionObserver is largely replaceable with Events (at a performance penalty).  Neither ResizeObserver nor IndexDBObserver can be implemented in terms of Events.

So in most cases, saying that "by default, you should use events; only use XObservers as a last resort," is a bit like saying, "By default, you should use the fetch API; only use getBoundingClientRect as a last resort."

**Comment by atotic** [comment]
@domenic "Observers should be used as a last resort" is not helpful to a spec designer. Enumerating strengths and drawbacks of both approaches would be very helpful.

When I started the design of RO, I initially designed it as an EventTarget. I got immediate pushback because of the alleged poor performance of events. I was eventually convinced that observer is the way to go, but it was not obvious. The arguments for observer were:
- event propagation is slow, and RO cannot afford the cost.
- event dispatch: we do not care about every resize, just the final resize before rAF.

Implementing RO as an event listener would have meant worse performance, and breaking developer expectation of event propagation. I never saw actual performance data on how much worse listeners are than observers.

In general, I like events for their established API. But they have other drawbacks too:
- event dispatch varies a lot between browsers. When one event will be dispatched wrt different event type is usually something to be answered empirically, not by reading the spec.
- no way to pass in custom options when adding listeners.

**Comment by domenic** [comment]
I think without performance data we should discount such pushback. I'm disappointed you got such pushback without data to back it up; hopefully by editing this document appropriately we can ensure such things don't happen again.

Event dispatch should not vary between browsers. If it does that is a spec bug. There is nothing inherently different about XObsever vs. EventTarget that makes it easier or harder to write imprecise specs with one versus the other.

I think that leaves the custom options as the reason to use XObserver. They should not be used unless you need that specific functionality. That is what I meant by "last resort", which I agree is imprecise phrasing.

**Comment by atotic** [comment]
Off topic, FYI: I did an [analysis](https://cdn.rawgit.com/atotic/event-loop/caa3cfd4/rendering-events.html) of event dispatch between browsers. There were a lot of differences.

**Comment by tabatkins** [comment]
I agree with Domenic that the core difference between XObserver and Events is that XObserver gives you more up-front control about the type/timing of information you receive, while Events are you "tapping in" to an existing information stream.

Theoretically we could offer more custom ways to begin listening to events, rather than just tapping the "firehose" of `addEventListener`, and we should get the same benefits.  That's more or less all that an XObserver *is*, tho, so it would be largely semantic.  The big benefit of doing that, I think, is what Domenic mentioned up above - events are a standardized interface that other tooling can already interact with in a generic fashion, while XObservers are always custom right now.

So, for example, we could have had a `IntersectionEventFilterer(el, {options}).addEventListener(cb)` sort of API. That gives us the common Event API, *plus* the filtering and some of the timing effects that an observer brings.

---

The other big difference, only relevant for some things, is timing, mentioned by a few others here. Events require you to return to the event loop, and thus happen at task speed; observers can happen at microtask speed.  I don't know if we can/should try to retrofit this capability onto events. If we don't, tho, it'll continue to drive the use of XObservers when this is relevant.

**Comment by smaug----** [comment]
Events don't require returning to the event loop.

**Comment by tabatkins** [comment]
Ah, I was under the impression they did! More evidence that we need a good explainer for all this that lays out the actual abilities and pros/cons, then. ^_^

**Comment by smaug----** [comment]
MutationObserver could have been 
var mo = new MutationObserver();
mo.onmutation = function(e) { /* do stuff with e.mutations;*/ }
mo.observe(...);

But that would have added the complication to support multiple listeners, and onmutation or addEventListener("mutation", ...) is just overhead without too good reasons.

Note, event propagation in plain simple EventTarget doesn't really cost anything, and calling either event listener or just some random callback should be about equally fast or slow.
Event propagation may show up in performance profiles when event has some path with length > 1.

**Comment by annevk** [comment]
It seems there is still massive confusion here. Events are not per definition costly. Some events have been defined/designed poorly (e.g., `scroll`), but you cannot generalize from that at all. Again, if you don't understand that events and callbacks are effectively identical and any observer can be done as an `EventTarget` object as @smaug---- shows above there's more learning to be done.

Therefore it's a good idea to start from a number of use cases and (performance) requirements and work your way towards an API from those. And as long as we don't quite understand the tradeoffs in enough detail to write them down it would do the platform good if you sought wide review for any addition.

**Comment by annevk** [comment]
@atotic did you end up filing browser bugs for the differences and land your test as a web-platform-tests test? Seems like that would be very good. (Firefox has a known issue with promises and microtasks that we hope to address shortly.)

**Comment by jakearchibald** [comment]
@smaug---- 

> But that would have added the complication to support multiple listeners, and onmutation

Where does this complication live? It's pretty easy from a spec point of view, and I would have assumed that implementing an event is easier than an XObserver, given that plumbing exists for the former, and the latter involves creating additional interfaces.

**Comment by smaug----** [comment]
The complication lives in the API definition. What if one listener calls take*Records() or causes more mutations etc. Without multiple callbacks, the observer is, how to say, self-contained.

**Comment by atotic** [comment]
@annevk I've filed browser bugs for Promise differences, but not others. Promise delivery on microtask queue is well specified, and I could write an actionable bug report.

I did not file bugs for the rest of the misdelivered events. I prefer writing bug reports that are more specific than "your event delivery order is borked". There is no direct js visibility into the event loop, and with multiple events out of order, I was unable to clearly attribute the failures: was timeout firing too late, or scroll & matchMedia too early, etc.

I'd be happy to convert my test page into w-p-t if that would help, just tell me where should I place the test.

**Comment by dmurph** [comment]
So I think I'm feeling the following on XObserver w/ callback vs Events, and why I think callbacks can be advantageous for Observers:

1. Callbacks allow it to be called as a microtask.
2. Callbacks allow much finer grained constraints on state. Observers have to reason about timing, where we are notifying at exact states (observing database changes, needing to read a consistant state, etc). An event model needs to allow for adding new listeners at any time, which makes this more complicated. There's a snapshot delta system happening in all of the XObserver cases, and having the event listener(s) being added at arbitrary times makes this more complex than it should be. It's harder to guarantee an initial state here for listeners.
3. Event models need to support event features like deferring / prevent propagation of events. This complicates things more for no reason.

So because of these problems I think callbacks are better for the XObserver cases where this is a snapshot-delta system (which I think is all of them).

**Comment by domenic** [comment]
> Callbacks allow it to be called as a microtask.

Events can do this too.

> An event model needs to allow for adding new listeners at any time

This is a great point I hadn't seen yet in this thread.

> Event models need to support event features like deferring / prevent propagation of events. This complicates things more for no reason.

These features are optional (events are by default non-cancelable and non-bubbling) so I don't find this argument persuasive.

**Comment by annevk** [comment]
To further illustrate @domenic's reply above, https://dom.spec.whatwg.org/#notify-mutation-observers handles invocation of both the mutation observer callback and firing of the `slotchange` event in the same microtask.

**Comment by annevk** [comment]
@atotic thanks, I think we have coverage for promise delivery already. For the remainder it's probably worth filing specification bugs if the timing relative to other events is unclear.

**Comment by domenic** [comment]
Thanks for the great discussion, all. I've tried to collate everything into https://github.com/w3ctag/design-principles/pull/79 , which I am pretty happy with. Review appreciated.

However, I am increasingly concerned that ResizeObserver made a mistake in its design. It doesn't seem to be taking advantage of any of the things I listed as advantages for XObserver. I will open an issue on that repository to discuss.

**Comment by ojanvafai** [comment]
Not sure whether to leave comments here or on the patch.

I'm not sure what you mean by "heavyweight" with respect to XObservers. Do you mean in terms of performance or ergonomics? If performance, I don't think that's quite right.

XObserver pattern is marginally more expensive for setting up the observer, but then cheaper for each individual dispatch. This is especially true for events on Nodes where you have to compute the ancestor chain up front for each event dispatch.

When we last looked into event dispatch performance being too high overhead, 90+% of it went into computing the ancestor chain. The problems is that computing is it O(n) in the depth of the tree, so what seems like a cheap dispatch now needs to do a bunch of tree walking. I can't seem to find any of the data from when we looked into it.

TL;DR: I would actually describe *events* as more heavyweight (on Nodes anyways).

**Comment by domenic** [comment]
> Do you mean in terms of performance or ergonomics?

Ergnomics, specifically with regard to the conceptual overhead and all the new types, as the rest of the paragraph discusses.

> This is especially true for events on Nodes where you have to compute the ancestor chain up front for each event dispatch.

As pointed out, you don't have to participate in the node hierarchy when firing events.

**Comment by ojanvafai** [comment]
I see. I think I misunderstood what this discussion was about suggesting that we should, e.g. fire a "resize" event on Nodes. 

The patch you have is a little confusing. It seems to bar arguing against the new object+constructor+dictionary, but then the IntersectionObserver example has all of those. So, I don't really see the ETIntersectionObserver example as lighter weight, but I do see that it naturally has all the other methods on EventTarget like "once", which seems good.

The IntersectionObserver as an EventTarget example in the pull request was very clarifying though. You're saying that instead of having observe/unobserve, we should have addEventListener/removeEventListener, and all the other things on EventTarget. That's easier for me to wrap my head around.

In fact, you could imagine that we have an EventTarget pattern for Observers that avoids the need for having a distinction at all:
 - Add disconnect() to EventTarget
 - Add dispatchPending() (equivalent to takeRecords()) to EventTarget
 - Allow for subclasses of EventTarget to take subclasses of EventListenerOptions as the third argument to addEventListener. This gets the ability to pass options to addEventListener instead of to the observe method.
